// @airswap/types/build/index.js
"use strict";
/*
  Copyright 2020 Swap Holdings Ltd.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("@airswap/constants");
exports.EIP712 = {
    EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'verifyingContract', type: 'address' },
    ],
    Order: [
        { name: 'nonce', type: 'uint256' },
        { name: 'expiry', type: 'uint256' },
        { name: 'signer', type: 'Party' },
        { name: 'sender', type: 'Party' },
        { name: 'affiliate', type: 'Party' },
    ],
    Party: [
        { name: 'kind', type: 'bytes4' },
        { name: 'wallet', type: 'address' },
        { name: 'token', type: 'address' },
        { name: 'amount', type: 'uint256' },
        { name: 'id', type: 'uint256' },
    ],
};
exports.emptyQuoteParty = {
    kind: constants_1.tokenKinds.ERC20,
    token: constants_1.ADDRESS_ZERO,
    amount: '0',
    id: '0',
};
exports.emptyOrderParty = {
    kind: constants_1.tokenKinds.ERC20,
    wallet: constants_1.ADDRESS_ZERO,
    token: constants_1.ADDRESS_ZERO,
    amount: '0',
    id: '0',
};
exports.emptySignature = {
    version: constants_1.signatureTypes.PERSONAL_SIGN,
    signatory: constants_1.ADDRESS_ZERO,
    validator: constants_1.ADDRESS_ZERO,
    r: constants_1.LOCATOR_ZERO,
    s: constants_1.LOCATOR_ZERO,
    v: '0',
};
